### ISSUE TITLE
Disable VHost when host does not support vhost-net

### ISSUE THREAD

**Issue by egernst** [kata-containers/runtime#169]
_From @sameo on September 27, 2017 9:48_

We enable `vhost-net` unconditionally:

https://github.com/containers/virtcontainers/blob/master/qemu.go#L292

We should dynamically toggle it depending on the host `vhost-net` availability.

_Copied from original issue: containers/virtcontainers#385_

**Comment by egernst** [comment]
_From @sameo on September 27, 2017 9:49_

cc @mcastelino

**Comment by egernst** [comment]
I'll take a look and see if this is still relevant...

**Comment by WeiZhang555** [comment]
I think this can be done by trying to insmod vhost-net kernel module, if ko can't be installed, the qemu vhost can be disabled.

**Comment by caoruidong** [comment]
I wonder if it is reasonable run kata-runtime after kata-check fails when there is no vhost_net module on the host?
https://github.com/kata-containers/runtime/blob/ed1e343b93d3822a36a01e672252ccc6da3389e0/cli/kata-check_amd64.go#L81-L83

**Comment by jodh-intel** [comment]
@caoruidong - I'm not sure I understand exactly what you are saying. However, the only way for the `kata-check` logic to be called is by running the CLI command - the runtime itself never uses it internally and that's by design since:

- It takes times to run those checks (bad :)
- The checks would end up being called multiple times for a standard `docker run -ti busybox true` (inefficient and redundant).
- The checks should never need to be run as part of normal operations as the runtime itself assumes the system is running on a system capable of exercising the main functionality.

In summary, `kata-check` and `kata-env` are purely "user/admin commands".

**Comment by caoruidong** [comment]
If host doesn't have vhost_net, kata-check fails which means system cannot run kata containers.
> [root@localhost vhost]# kata-runtime kata-check
INFO[0000] CPU property found                            arch=amd64 description="Intel Architecture CPU" name=GenuineIntel pid=21811 source=runtime type=attribute
INFO[0000] CPU property found                            arch=amd64 description="Virtualization support" name=vmx pid=21811 source=runtime type=flag
INFO[0000] CPU property found                            arch=amd64 description="64Bit CPU" name=lm pid=21811 source=runtime type=flag
INFO[0000] CPU property found                            arch=amd64 description=SSE4.1 name=sse4_1 pid=21811 source=runtime type=flag
INFO[0000] kernel property found                         arch=amd64 description="Kernel-based Virtual Machine" name=kvm pid=21811 source=runtime type=module
INFO[0000] kernel property found                         arch=amd64 description="Intel KVM" name=kvm_intel pid=21811 source=runtime type=module
WARN[0000] kernel module parameter has unexpected value  arch=amd64 description="Intel KVM" expected=Y name=kvm_intel parameter=nested pid=21811 source=runtime type=module value=N
INFO[0000] Kernel property value correct                 arch=amd64 description="Intel KVM" expected=Y name=kvm_intel parameter=unrestricted_guest pid=21811 source=runtime type=module value=Y
INFO[0000] kernel property found                         arch=amd64 description="Host kernel accelerator for virtio" name=vhost pid=21811 source=runtime type=module
ERRO[0000] kernel property not found                     arch=amd64 description="Host kernel accelerator for virtio network" name=vhost_net pid=21811 source=runtime type=module
ERRO[0000] ERROR: System is not capable of running Kata Containers  arch=amd64 name=kata-runtime pid=21811 source=runtime
ERROR: System is not capable of running Kata Containers

But if kata can run with vhost_net disabled, is it necessary to check vhost_net?

**Comment by mcastelino** [comment]
@egernst we should allow a non vhost device setup. Preferably via the configuration vs runtime detection. So default enable unless configuration tells us not to. Auto detect is probably not the best way to go about doing it.

**Comment by egernst** [comment]
@mcastelino - do we have a ballpark on what perf gains /losses this provides when we have and don't have vhost-net?
